### PR TITLE
SKALE-4691 Update Schain structure with generation field

### PR DESCRIPTION
**Pull request by DimaStebaev**

### CODE REVIEW

File: scripts/test_deploy.sh [hunk 1]
```diff
@@ -1,6 +1,8 @@
 #!/usr/bin/env bash
 
+set -e
+
 npx ganache-cli --gasLimit 8000000 --quiet &
 GANACHE_PID=$!
-PRODUCTION=true npx hardhat run migrations/deploy.ts --network localhost || exit $?
+PRODUCTION=true npx hardhat run migrations/deploy.ts --network localhost
```
Comment (payvint) [inline scripts/test_deploy.sh:7]: Why did you remove exit? I think it should not be removed - if deploy fails - build fails

Comment (DimaStebaev) [reply scripts/test_deploy.sh:7]: I added `set -e` command above. It setups bash to exit a script with non zero if any command inside fails.

Comment (payvint) [reply scripts/test_deploy.sh:7]: Good

**Comment by payvint** [comment]
How we would initialize generation in already created schains?
And are you sure that upgrade would be smoothly done if we just add variable in the struct.
I remember that we have almost the same with domainNames. And it was failed with adding it to the struct - but there was an array

**Comment by DimaStebaev** [comment]
All existing schains will have 0 generation because the new field is not initialized.
I added new field to the structure end so there are no problems with memory.
The problem with domain names is not relevant because there were array of structs. They are stored sequentially in the storage and modification of struct length causes shifting of fields location.
In current case `Schain` structure is a value in a mapping and each value is stored in separate place in storage what means that the length modification does not affects other stored structures.

**Comment by payvint** [comment]
> All existing schains will have 0 generation because the new field is not initialized.
> I added new field to the structure end so there are no problems with memory.
> The problem with domain names is not relevant because there were array of structs. They are stored sequentially in the storage and modification of struct length causes shifting of fields location.
> In current case `Schain` structure is a value in a mapping and each value is stored in separate place in storage what means that the length modification does not affects other stored structures.

Got it - good explanation - thanks!

File: contracts/SchainsInternal.sol [hunk 2]
```diff
@@ -546,6 +565,13 @@ contract SchainsInternal is Permissions, ISchainsInternal {
         return (schainTypes[typeOfSchain].partOfNode, schainTypes[typeOfSchain].numberOfNodes);
     }
 
+    /**
+     * @dev Returns generation of a particular schain
+     */
+    function getGeneration(bytes32 schainHash) external view returns (uint) {
```
Comment (payvint) [inline contracts/SchainsInternal.sol]: If there is no schain this function will return a 0 generation, which is a real generation of some schains. This function may confuse users.

File: contracts/ConstantsHolder.sol [hunk 3]
```diff
@@ -80,7 +80,7 @@ contract ConstantsHolder is Permissions {
     uint public constant BROADCAST_DELTA = 131000;
     uint public constant COMPLAINT_BAD_DATA_DELTA = 49580;
     uint public constant PRE_RESPONSE_DELTA = 74500;
-    uint public constant COMPLAINT_DELTA = 83069;
+    uint public constant COMPLAINT_DELTA = 86060;
```
Comment (payvint) [inline contracts/ConstantsHolder.sol:83]: Why did you change complaint delta? Was there some changes in gas consumption changes of some operations? Or here was an engineering debt of London change? Are you sure that every other delta is correct?

Comment (DimaStebaev) [reply contracts/ConstantsHolder.sol:83]: I'm not sure. Is there a way to calculate these constants more precise?
I've found that our tests fail and have increased for minimum value to satisfy tests.
https://github.com/skalenetwork/skale-manager/runs/4236923596?check_suite_focus=true

File: contracts/SchainsInternal.sol (hunk 2 above)
Comment (DimaStebaev) [reply contracts/SchainsInternal.sol]: Fixed.

**Review by payvint — APPROVED** [review]
Looks good!

**Review by yavrsky — APPROVED** [review]
